StreamReader.cpp

```diff
@@ -20,6 +20,11 @@ FileStreamReader::~FileStreamReader() {
 }
 
 void FileStreamReader::Read(char* buffer, size_t size) {
+	// Unable to add a std::streampos to a size_t without casting.
+	if (static_cast<size_t>(file.tellg()) + size > Length()) {
```

Comment:
This looks correct, but causes seeking for each Read due to the implementation of `Length`. There's a potential for a large performance impact here, especially if the seeking affects buffering, such as by clearing read-ahead buffers.

It may be more efficient to seek back to the original position in the case of a failed or partial read than to pre-check.

It might also make sense to make no guarantees about the file pointer or buffer contents after a failed or partial read. This is the issue of exception safety. No undefined behaviour has occurred yet, but the object may have been left in an unknown state, making future calls to certain methods potentially unsafe. Not all methods though, as object destruction must necessarily still be safe.

It's nice to avoid undefined behaviour when possible. As such, I think I'd recommend pre-recording the position, wrapping the read in a try/catch, where the catch restores the file position and rethrows the same exception.

Reply:
Agreed there is a performance issue. I was avoiding recording the length of the file separately in case the file size changed mid read. While this makes sense in the StreamWriter, it doesn't here because the file should not be changing size during a read only operation. I'll add the pre-recorded length.

I may wait to work on improved exception handling until a nother branch. Partially because I don't understand how std::ifstream itself throws exceptions/notifies of errors.

Reply:
Correct, the file can change size. Even if it's a read only stream, another process on the system might change the size of the file. Any set of check+do operations which are not atomic will always be subject to race conditions, and so can never ensure correctness. This is why it's best to react to the result of the read, rather than trying to pre-check it. 

I was thinking record the current position at the beginning of the read, and if there is an error during the read, to catch it and seek to the previously recorded position and then (re)throw an error. The performance impact is less, only a read of the current position for each call to Read, and only seeking to recover the internal stream state after an error.

Of course, now that I'm thinking about it, even that would be a non-atomic set of operations. If the file size changed because it was truncated to 0, the seek would not be very meaningful.

This is probably why some APIs choose to avoid the performance impact of trying to make guarantees under error conditions. Sometimes you can't really make a meaningful guarantee, so why bother take the performance hit. Just document what can and can't be relied upon during those circumstances.

If you wanted to go with no guarantee in the case of exceptions, I'd be willing to go with that. We'd just need to document it.